src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+            return WRITE;
+        case OS_READ_ONLY :
+            return READ;
+        default :
+            return NONE;
+    }
+    PANIC("unreachable statement");
+    return 0;
+}
+
+uint32 permission_cos_to_cFE(permission_t permission) {
+    switch (permission) {
+        case READ_WRITE :
+            return OS_READ_WRITE;
+        case WRITE :
+            return OS_WRITE_ONLY;
+        case READ :
+            return OS_READ_ONLY;
+        default :
+            PANIC("Invalid permission in existing file");
+    }
+    PANIC("unreachable statement");
+    return 0;
+}
+
+// checks if a path is the right format for a path, and if it exists
+int32 path_chk_exists(const char *path)
+{
+    int32 ret = path_chk_isvalid(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    if (file_find((char *)path) == NULL) return OS_FS_ERROR;
+    return 0;
+}
+
+// checks if a path is the correct format, for example to be inserted
+int32 path_chk_isvalid(const char *path)
+{
+    if (path == NULL) return OS_FS_ERR_INVALID_POINTER;
+    if (strlen(path) > OS_MAX_PATH_LEN) return OS_FS_ERR_PATH_TOO_LONG;
+    if (strlen(path_to_name((char *)path)) > OS_MAX_FILE_NAME) return OS_FS_ERR_NAME_TOO_LONG;
+    if (path[0] != '/') return OS_FS_ERR_PATH_INVALID;
+    assert(openfs);
+
+    if (openfs->root) {
+        assert(openfs->root->name);
+        if (memcmp(openfs->root->name, path + 1, strlen(openfs->root->name))) {
+            return OS_FS_ERR_PATH_INVALID;
+        }
     }
     return OS_FS_SUCCESS;
 }
 
-//this could be more performant bithack
-uint32 round_to_blocksize(uint32 offset)
+int32 path_translate(char *virt, char *local)
 {
-    if (offset % TAR_BLOCKSIZE) return offset + (TAR_BLOCKSIZE - (offset % TAR_BLOCKSIZE));
-    return offset;
+    if (!virt || !local) return OS_FS_ERR_INVALID_POINTER;
+    int32 ret = path_chk_isvalid(virt);
+    if (ret != OS_FS_SUCCESS) return ret;
+    if (!openfs->root) return OS_FS_ERR_PATH_INVALID;
+    ret = path_chk_exists(virt);
+    if (ret != OS_FS_SUCCESS) return ret;
+    strcpy(local, virt);
+    return OS_FS_SUCCESS;
 }
 
 /******************************************************************************
-** fsobj Level Methods
+** f_part Level Methods
 ******************************************************************************/
 
-uint32 fsobj_init(struct fsobj *o)
+uint32 newpart_get(struct f_part **part)
 {
-    assert(o);
-    assert(o->filedes > 0);
+    if (!ci) {
+        struct cos_defcompinfo *defci = cos_defcompinfo_curr_get();
+        ci = &defci->ci;
+    }
+
+    *part = cos_page_bump_alloc(ci);
 
-    o->name     = "EXAMPLE_NAME";
-    o->type     = FSOBJ_DIR;
-    o->size     = 0;
-    o->refcnt   = 0;
-    o->data     = NULL;
-    o->child    = o->parent = NULL;
-    o->next     = o->prev = NULL;
+    assert(part != NULL);
+    (*part)->next = NULL;
+    (*part)->prev = NULL;
+    (*part)->file = NULL;
+    (*part)->data = (char *) *part + sizeof(struct f_part);
```

Comment:
flexable array member cannot work in this case, as the data in a static file (TAR loaded file) is not sequentially right after the f_part struct for that file